### PR TITLE
New caching design/logic

### PR DESCRIPTION
**Pull request by andersfylling**
The current caching system is difficult to maintain and read. This PR hopes to resolve that. Simplicity over performance is favored as this can be improved later on.

This also introduces fake/local events created by disgord and not discord. However, these should be isolated by default and users can gain access to them through a config option.

Cache should be structured into REST and event calls; where event calls starts with on<EventName> and the REST calls ??????.

Since some events and perhaps REST calls do not yield complete data structures, the cache layer needs to deal with the raw JSON. This allows partial updating and since the raw json is a read only value, each of the different cache repository can work on the data in parallel.

### Events
 - [x] Ready
       - cache disabled / enabled
       - init unavailable guilds
       - update/store bot user data
 - [x] Resumed
 - [ ] ChannelCreate
       - cache disabled / enabled
       - cache users from recipient list
       - add to guild if guild_id is set
       -! _**Do or do not overwrite existing entry?**_
 - [x] ChannelUpdate
       - cache disabled / enabled
       - cache users from recipient list
       - make sure it exists in guild if guild_id is set
       - update existing entry
 - [x] ChannelDelete
       - cache disabled / enabled
       - delete entry if it exists
       - update guild channel id slice
 - [ ] ChannelPinsUpdate
       - cache disabled / enabled
       - update related channel & store messages in channel
 - [x] TypingStart
       - cache disabled
 - [x] MessageCreate
       - cache disabled
       - update related channel: last msg id, if id unix is same or newer
 - [x] MessageUpdate
       - cache disabled
 - [x] MessageDelete
       - cache disabled
 - [x] MessageDeleteBulk
       - cache disabled
 - [ ] MessageReactionAdd
       - cache disabled / enabled
 - [ ] MessageReactionRemove
       - cache disabled / enabled
 - [ ] MessageReactionRemoveAll
       - cache disabled / enabled
 - [ ] GuildEmojisUpdate
       - cache disabled / enabled
 - [ ] GuildCreate
       - cache disabled / enabled
 - [ ] GuildUpdate
       - cache disabled / enabled
 - [ ] GuildDelete
       - cache disabled / enabled
 - [ ] GuildBanAdd
       - cache disabled / enabled
 - [ ] GuildBanRemove
       - cache disabled / enabled
 - [ ] GuildIntegrationsUpdate
       - cache disabled / enabled
 - [ ] GuildMemberAdd
       - cache disabled / enabled
 - [ ] GuildMemberRemove
       - cache disabled / enabled
 - [ ] GuildMemberUpdate
       - cache disabled / enabled
 - [ ] GuildMembersChunk
       - cache disabled / enabled
 - [ ] GuildRoleCreate
       - cache disabled / enabled
 - [ ] GuildRoleUpdate
       - cache disabled / enabled
 - [ ] GuildRoleDelete
       - cache disabled / enabled
 - [ ] PresenceUpdate
       - cache disabled / enabled
 - [x] UserUpdate
       - cache disabled / enabled
 - [ ] VoiceStateUpdate
       - cache disabled / enabled
 - [ ] VoiceServerUpdate
       - cache disabled / enabled
 - [x] WebhooksUpdate
       - cache disabled / enabled

### CODE REVIEW

File: cache/lfu/cache.go [hunk 1]
```diff
@@ -186,6 +189,23 @@ func (list *CacheList) Delete(id Snowflake) {
...
 
+func (list *CacheList) Foreach(cb func(interface{})) {
+	for k := range list.items {
+		cb(list.items[k])
+	}
+}
+
+func (list *CacheList) ListIDs() (ids []snowflake.ID) {
+	ids = make([]snowflake.ID, len(list.items))
+	var i int
```
Comment (mlvzk) [inline cache/lfu/cache.go]: isn't the `i` useless here? the `k` below could be used instead to index `ids`

File: cache/lru/cache.go [hunk 2]
```diff
@@ -150,6 +155,23 @@ func (list *CacheList) Delete(id Snowflake) {
...
 
+func (list *CacheList) Foreach(cb func(interface{})) {
+	for k := range list.items {
+		cb(list.items[k])
+	}
+}
+
+func (list *CacheList) ListIDs() (ids []snowflake.ID) {
+	ids = make([]snowflake.ID, len(list.items))
+	var i int
```
Comment (mlvzk) [inline cache/lru/cache.go]: same as before
> isn't the `i` useless here? the `k` below could be used instead to index `ids`

File: cache/tlru/cache.go [hunk 3]
```diff
@@ -166,6 +169,23 @@ func (list *CacheList) Delete(id Snowflake) {
...
 
+func (list *CacheList) Foreach(cb func(interface{})) {
+	for k := range list.items {
+		cb(list.items[k])
+	}
+}
+
+func (list *CacheList) ListIDs() (ids []snowflake.ID) {
+	ids = make([]snowflake.ID, len(list.items))
+	var i int
```
Comment (mlvzk) [inline cache/tlru/cache.go]: same as before
> isn't the `i` useless here? the `k` below could be used instead to index `ids`

File: cache/lfu/cache.go [hunk 4]
```diff
@@ -186,6 +189,23 @@ func (list *CacheList) Delete(id Snowflake) {
 	}
 }
 
+func (list *CacheList) Foreach(cb func(interface{})) {
```
[the remark this answers is not in the document]
Comment (andersfylling) [reply cache/lfu/cache.go]: Reason this is left alone, for now. Is to allow for a single interface. Tbh I have decided that the best caching strategy is to use TLRU as discussed in this paper: https://www.researchgate.net/publication/260255765_Time_Aware_Least_Recent_Used_TLRU_cache_management_policy_in_ICN

**Comment by andersfylling** [comment]
moved to v0.13 due to amount of work

**Comment by andersfylling** [comment]
Each repo must handle enable/disable caching on it's own.

**Comment by andersfylling** [comment]
The waitgroup use can be hidden inside a builder object or something to clean up the code.

**Comment by andersfylling** [comment]
The current cache interface looks like:
```go


type Cache interface {
	// Clear clears up the cache and removes temporary configuration/dynamic settings that are unique per session,
	// if any.
	Clear(repository cacheRepository)

	Lock(repository cacheRepository)
	Unlock(repository cacheRepository)
	RLock(repository cacheRepository)
	RUnlock(repository cacheRepository)

	// Del* deletes an entry
	DelUser(id Snowflake)
	DelChannel(id Snowflake)
	DelGuild(id Snowflake)
	DelPresence(id Snowflake)

	GetUser(id Snowflake) (*User, bool)
	GetChannel(id Snowflake) (*Channel, bool)
	GetGuild(id Snowflake) (*Guild, bool)
	GetPresence(guildID, userID Snowflake) (*UserPresence, bool)

	SetUser(id Snowflake, user *User, cost int64) bool
	SetChannel(id Snowflake, channel *Channel, cost int64) bool
	SetGuild(id Snowflake, guild *Guild, cost int64) bool
	SetPresence(guildID, userID Snowflake, presence *UserPresence, cost int64) bool
}
```

if there are any concerns about this, please comment.

**Pull request closed without merging**